content/blog/2017/05/2017-05-12-kubernetes-journey-on-azure.adoc

```diff
@@ -0,0 +1,397 @@
+ |     Azure:  |                                  |            └── service.yaml    |
+ | K8s Cluster |                                  |________________________________|
+ |_____________|
+
+----
+
+The main benefit of this approach, is to let puppet manage the environment and
+run common tasks.  And if needed, we still have a place where we can go to run
+uncommon operations.
+
+In following example, we define one puppet class for Datadog.
+
+.Puppet class for resource Datadog
+[source,puppet]
+----
+# Deploy datadog resources on kubernetes cluster
+#   Class: profile::kubernetes::resources::datadog
+#
+#   This class deploy a datadog agent on each kubernetes node
+#
+#   Parameters:
+#     $apiKey:
+#       Contain datadog api key.
+#       Used in secret template
+class profile::kubernetes::resources::datadog (
+    $apiKey = base64('encode', $::datadog_agent::api_key, 'strict')
+  ){
+  include ::stdlib
+  include profile::kubernetes::params
+  require profile::kubernetes::kubectl
+
+  file { "${profile::kubernetes::params::resources}/datadog":
+    ensure => 'directory',
+    owner  => $profile::kubernetes::params::user,
+  }
+
+  profile::kubernetes::apply { 'datadog/secret.yaml':
+    parameters => {
+        'apiKey' => $apiKey
+    },
+  }
+  profile::kubernetes::apply { 'datadog/daemonset.yaml':}
+  profile::kubernetes::apply { 'datadog/deployment.yaml':}
+
+  # As secret's changes do not trigger pods update,
+  # we must reload pods 'manually' in order to use updated secrets.
+  # If we delete a pod defined by a daemonset,
+  # this daemonset will recreate pods automatically.
+  exec { 'Reload datadog pods':
+    path        => ["${profile::kubernetes::params::bin}/"],
+    command     => 'kubectl delete pods -l app=datadog',
+    refreshonly => true,
+    environment => ["KUBECONFIG=${profile::kubernetes::params::home}/.kube/config"] ,
+    logoutput   => true,
+    subscribe   => [
+      Exec['apply datadog/secret.yaml'],
+      Exec['apply datadog/daemonset.yaml'],
+    ],
+  }
+}
+----
+->
+https://github.com/jenkins-infra/jenkins-infra/tree/staging/dist/profile/manifests/kubernetes/resources[More "resources" examples]
+
+
+Let's compare the Puppet approach with the challenges discovered with the
+Jenkins approach.
+
+=== Order matters
+
+With Puppet, it becomes easier to define priorities as
+puppet provides relationship meta parameters and the function 'require' (see
+also:
+https://docs.puppet.com/puppet/4.9/lang_relationships.html[Puppet relationships])
+
+In our Datadog example, we are sure that deployment will respect following order:
+
+  datadog/secret.yaml -> datadog/daemonset.yaml -> datadog/deployment.yaml
+
+Currently, our Puppet code only applies configuration when it detects files’
+changes.  It would be better to compare local files with the cluster
+configurations in order to trigger required updates, but we haven't found a
+good way to improve this yet.
```

Comment:
Improve what? Shouldn't this be e.g. "implement", as two approaches are contrasted here, and the latter is being brought up as better?